gubernator/github/periodic_sync.py

```diff
@@ -0,0 +1,171 @@
+    if write_html is None:
+        write_html = lambda x: None
+
+    logging.info('syncing repo %s', repo)
+    write_html('<h1>%s</h1>' % repo)
+
+    # There is a race condition here:
+    # We can't atomically get a list of PRs from the database and GitHub,
+    # so a PR might falsely be in stale_open_prs if it is opened after
+    # we scan GitHub, or falsely be in missing_prs if a PR is made after we
+    # got the list from GitHub, and before we get the list from the database.
+    #
+    # These cases will both be fixed the next time this code runs, so we don't
+    # try to prevent it here.
+    prs_gh = get_prs_from_github(token, repo)
+    prs_gh_by_number = {pr['number']: pr for pr in prs_gh}
+
+    prs_db = list(models.GHIssueDigest.find_open_prs_for_repo(repo))
+    prs_db_by_number = {pr.number: pr for pr in prs_db}
+
+    numbers_datastore = set(prs_db_by_number)
+    numbers_github = set(prs_gh_by_number)
+
+    stale_open_prs = sorted(numbers_datastore - numbers_github)
+    missing_prs = sorted(numbers_github - numbers_datastore)
+
+    if not stale_open_prs and not missing_prs:
+        write_html('matched, no further work needed')
+        logging.info('matched, no further work needed')
+        return
+
+    logging.info('PRs to close: %s', stale_open_prs)
+    logging.info('PRs to open: %s', missing_prs)
+
+    write_html('<br>')
+    write_html('PRs that should be closed: %s<br>' % stale_open_prs)
+
+    for number in stale_open_prs:
+        pr = prs_db_by_number[number]
+        write_html('<b>%d</b><br>%s<br>' % (number, pr))
+        inject_event_and_reclassify(repo, number, 'gh-sync-close',
+            {'state': 'closed',
+             # These other 3 keys are injected because the classifier expects them.
+             # This simplifies the testing code, and means we don't have .
+             'user': {'login': pr.payload['author']},
+             'assignees': [{'login': u} for u in pr.payload['assignees']],
+             'title': pr.payload['title']})
+
+    write_html('PRs that should be opened: %s<br>' % missing_prs)
+
+    for number in missing_prs:
+        pr = models.shrink(prs_gh_by_number[number])
+        write_html('<br>%d</br><pre>%s</pre><br>' %
+            (number, json.dumps(pr, indent=4, sort_keys=True)))
+        inject_event_and_reclassify(repo, number, 'gh-sync', pr)
+
+
+class PRSync(webapp2.RequestHandler):
+    def get(self):
+        # This is called automatically by the periodic cron scheduler.
+        # For debugging, visit something like /sync?repo=kubernetes/test-infra
+        token = secrets.get('github_token', per_host=False)
+        if not token:
+            logging.warning('no github token, skipping sync')
+            self.abort(200)
+
+        # first, determine which repositories we need to sync
+        open_prs = list(
+            models.GHIssueDigest.find_open_prs().fetch(keys_only=True))
+        open_repos = sorted({models.GHIssueDigest(key=pr).repo for pr in open_prs})
```

Comment:
You can move this logic into the `else` case below.

Reply:
it's for debugging, and I like having the list to copy-paste from :)